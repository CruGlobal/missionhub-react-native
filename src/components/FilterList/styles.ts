import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  pageContainer: {
    backgroundColor: theme.primaryColor,
  },
  list: {
    flex: 1,
    backgroundColor: theme.white,
  },
});
