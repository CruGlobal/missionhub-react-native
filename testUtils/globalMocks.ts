import faker from 'faker/locale/en';
import { IMocks } from 'graphql-tools';
import moment from 'moment';

import {
  CommunityCelebrationCelebrateableEnum,
  PermissionEnum,
  ReminderTypeEnum,
  StepTypeEnum,
} from '../__generated__/globalTypes';

let currentId = 1;
const nextId = () => currentId++;
export const resetGlobalMockSeeds = () => {
  currentId = 1;
  faker.seed(1);
};
resetGlobalMockSeeds();

export const globalMocks: IMocks = {
  String: () => faker.lorem.words(),
  Int: () => faker.random.number(),
  Float: () => faker.random.number({ precision: 0.01 }),
  Boolean: () => faker.random.boolean(),
  ID: () => nextId(),
  ISO8601DateTime: () => faker.date.past(10, '2020-01-14').toISOString(),
  ISO8601Date: () =>
    moment(faker.date.past(10, '2020-01-14')).format('YYYY-MM-DD'),

  BasePageInfo: () => ({
    endCursor: null,
    hasNextPage: false,
    hasPreviousPage: false,
    startCursor: null,
  }),
  ReminderTypeEnum: () => ReminderTypeEnum.once,
  StepTypeEnum: () => faker.random.arrayElement(Object.values(StepTypeEnum)),

  Step: () => ({
    title: faker.lorem.sentence(),
  }),
  Person: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      fullName: `${firstName} ${lastName}`,
      firstName,
      lastName,
    };
  },
  Community: () => ({
    name: faker.company.catchPhrase(),
  }),
  CommunityCelebrationItem: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      celebrateableType: faker.random.arrayElement(
        Object.values(CommunityCelebrationCelebrateableEnum),
      ),
      changedAttributeValue: moment(
        faker.date.past(10, '2020-01-14'),
      ).toISOString(),
      subjectPersonName: `${firstName} ${lastName}`,
    };
  },
  CommunityPermission: () => {
    return {
      permission: faker.random.arrayElement(Object.values(PermissionEnum)),
    };
  },
  FeedItem: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      subjectPersonName: `${firstName} ${lastName}`,
    };
  },
};
