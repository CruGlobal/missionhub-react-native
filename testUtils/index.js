import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
