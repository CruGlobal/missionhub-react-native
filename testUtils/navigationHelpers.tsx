import React from 'react';
import {
  // eslint-disable-next-line import/named
  createAppContainer,
  // eslint-disable-next-line import/named
  createSwitchNavigator,
  // eslint-disable-next-line import/named
  NavigationParams,
  // eslint-disable-next-line import/named
  NavigationScreenProp,
} from 'react-navigation';
import { render } from 'react-native-testing-library';

export const createNavigationProp = (params?: NavigationParams) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let navigationProp: NavigationScreenProp<any> | undefined;

  const Navigator = createAppContainer(
    createSwitchNavigator({
      TestScreen: {
        screen: ({
          navigation,
        }: {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          navigation: NavigationScreenProp<any>;
        }) => {
          navigationProp = navigation;
          return null;
        },
        params,
      },
    }),
  );
  render(<Navigator />);

  if (!navigationProp) {
    throw 'Unable to get navigation screen prop';
  }

  return navigationProp;
};
