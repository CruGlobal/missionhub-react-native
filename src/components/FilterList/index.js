import React, { Component } from 'react';
import { View, ScrollView } from 'react-native';
import PropTypes from 'prop-types';

import Header from '../Header';
import BackButton from '../../containers/BackButton';
import FilterItem from '../FilterItem';

import styles from './styles';

export default class FilterList extends Component {
  handleDrillDown = item => {
    this.props.onDrillDown(item);
  };

  handleToggle = item => {
    this.props.onToggle(item);
  };

  render() {
    const { title, options, toggleOptions } = this.props;
    return (
      <View style={styles.pageContainer}>
        <Header left={<BackButton />} title={title} />
        <ScrollView style={{ flex: 1 }}>
          {options &&
            options.map(o => (
              <FilterItem
                key={o.id}
                item={o}
                onSelect={this.handleDrillDown}
                type="drilldown"
              />
            ))}
          {toggleOptions &&
            toggleOptions.map(o => (
              <FilterItem
                key={o.id}
                item={o}
                onSelect={this.handleToggle}
                type="switch"
                isSelected={o.selected}
              />
            ))}
        </ScrollView>
      </View>
    );
  }
}

FilterList.propTypes = {
  onDrillDown: PropTypes.func,
  onToggle: PropTypes.func,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      text: PropTypes.string,
      drilldown: PropTypes.array,
    }),
  ),
  toggleOptions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      selected: PropTypes.bool,
    }),
  ),
};
