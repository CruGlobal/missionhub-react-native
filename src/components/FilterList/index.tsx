import React from 'react';
import { ScrollView, View } from 'react-native';

import Header from '../Header';
import BackButton from '../../containers/BackButton';
import FilterItem from '../FilterItem';

import styles from './styles';

type FilterItem = object;

interface FilterListProps {
  onDrillDown: (item: FilterItem) => void;
  onToggle: (item: FilterItem) => void;
  title?: string;
  options?: {
    id?: string;
    text?: string;
    drilldown?: object[];
  }[];
  toggleOptions?: {
    id?: string;
    selected?: boolean;
  }[];
}

const FilterList = ({
  title,
  options,
  onDrillDown,
  onToggle,
  toggleOptions,
}: FilterListProps) => {
  const handleDrillDown = (item: object) => {
    onDrillDown(item);
  };
  const handleToggle = (item: object) => {
    onToggle(item);
  };
  return (
    <View style={styles.pageContainer}>
      <Header left={<BackButton />} title={title} />
      <ScrollView style={styles.list}>
        {options &&
          options.map((o, i) => (
            <FilterItem
              key={o.id}
              item={o}
              testID={`FilterListOption${i}`}
              onSelect={handleDrillDown}
              type="drilldown"
            />
          ))}
        {toggleOptions &&
          toggleOptions.map((o, i) => (
            <FilterItem
              key={o.id}
              item={o}
              testID={`FilterListToggleOption${i}`}
              onSelect={handleToggle}
              type="switch"
              isSelected={o.selected}
            />
          ))}
      </ScrollView>
    </View>
  );
};

export default FilterList;
