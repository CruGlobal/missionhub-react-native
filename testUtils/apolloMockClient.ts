import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { DocumentNode } from 'graphql';
import { buildClientSchema, IntrospectionQuery } from 'graphql/utilities';
import gql from 'fraql';
import { createMockerFromIntrospection } from 'fraql/mock';

import introspectionQuery from '../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks: IMocks = {}) => {
  const schema = buildClientSchema(
    (introspectionQuery as unknown) as IntrospectionQuery,
  );

  addMockFunctionsToSchema({
    schema,
    mocks: { ...globalMocks, ...mocks },
  });

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new SchemaLink({ schema }),
  });
};

export const mockFragment = <TData>(
  fragmentDocument: DocumentNode,
  options?: {
    mocks: IMocks;
  },
): TData => {
  if (!fragmentDocument.loc) {
    throw 'Unable to get original gql query to create FraQL fragment.';
  }

  return createMockerFromIntrospection(introspectionQuery, {
    mocks: globalMocks,
  }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
