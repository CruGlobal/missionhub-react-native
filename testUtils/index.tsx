import React, { ReactElement, ReactNode } from 'react';
import 'react-native';
import { Provider } from 'react-redux';
import { Provider as ProviderLegacy } from 'react-redux-legacy';
import thunk from 'redux-thunk';
import configureStore, { MockStore } from 'redux-mock-store';
// eslint-disable-next-line import/named
import { NavigationParams, NavigationProvider } from 'react-navigation';
import { ApolloProvider } from '@apollo/react-hooks';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderAPI } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';
import Enzyme, { shallow as enzymeShallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IMocks } from 'graphql-tools';

import { createApolloMockClient } from './apolloMockClient';
import { createNavigationProp } from './navigationHelpers';

Enzyme.configure({ adapter: new Adapter() });

export const createThunkStore = configureStore([thunk]);

interface RenderWithContextParams {
  initialState?: {};
  store?: MockStore;
  navParams?: NavigationParams;
  mocks?: IMocks;
  noWrappers?: boolean;
}

// Inspiration from https://github.com/kentcdodds/react-testing-library/blob/52575005579307bcfbe7fbe4ef4636147c03c6fb/examples/__tests__/react-redux.js#L69-L80
export function renderWithContext(
  component: ReactElement,
  {
    initialState,
    store = createThunkStore(initialState),
    navParams,
    mocks: mocks = {},
    noWrappers = false,
  }: RenderWithContextParams = {},
) {
  const mockApolloClient = createApolloMockClient(mocks);

  const navigation = createNavigationProp(navParams);

  let renderResult: RenderAPI;
  if (noWrappers) {
    renderResult = render(React.cloneElement(component, { navigation }));
  } else {
    // Warning: don't call any functions in here that return new instances on every call. All the props need to stay the same otherwise rerender won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
      <NavigationProvider value={navigation}>
        <ProviderLegacy store={store}>
          <Provider store={store}>
            <ApolloProvider client={mockApolloClient}>
              {children}
            </ApolloProvider>
          </Provider>
        </ProviderLegacy>
      </NavigationProvider>
    );

    renderResult = render(React.cloneElement(component, { navigation }), {
      wrapper,
    });
  }

  let storedSnapshot: ReactTestRendererJSON | null;
  return {
    ...renderResult,
    store,
    snapshot: () => expect(renderResult.toJSON()).toMatchSnapshot(),
    recordSnapshot: () => (storedSnapshot = renderResult.toJSON()),
    diffSnapshot: () => {
      if (!storedSnapshot) {
        throw new Error(
          'You must call recordSnapshot to store an initial snapshot before calling diffSnapshot',
        );
      }
      expect(
        snapshotDiff(storedSnapshot, renderResult.toJSON()),
      ).toMatchSnapshot();
    },
  };
}

// TODO: Remove all legacy rendering functions below

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (component: ReactElement) => {
  const { toJSON } = render(component);
  expect(toJSON()).toMatchSnapshot();
};

export const renderShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  let renderedComponent = enzymeShallow(
    <ProviderLegacy store={store}>{component}</ProviderLegacy>,
  ).dive();

  // If component has translation wrappers, dive deeper
  while (
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    ((renderedComponent.type() as any).displayName || '').startsWith(
      'withI18nextTranslation(',
    )
  ) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};
