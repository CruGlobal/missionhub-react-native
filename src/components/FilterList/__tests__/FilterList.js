import React from 'react';

import FilterList from '..';

import { renderShallow } from '../../../../testUtils';

const options = [{ id: '1', text: 'option1' }, { id: '2', text: 'option 2' }];
const toggleOptions = [
  { id: '3', selected: true },
  { id: '2', selected: false },
];
const onDrillDown = jest.fn();
const onToggle = jest.fn();

const component = renderShallow(
  <FilterList
    options={options}
    toggleOptions={toggleOptions}
    onDrillDown={onDrillDown}
    onToggle={onToggle}
  />,
);

it('renders correctly', () => {
  expect(component).toMatchSnapshot();
});

it('handles drilldown', () => {
  const drilldownItem = component
    .dive()
    .childAt(1)
    .childAt(0);
  const { item, type, onSelect } = drilldownItem.props();
  expect(type).toEqual('drilldown');

  onSelect(item);
  expect(onDrillDown).toHaveBeenCalledWith(item);
});

it('handles toggle', () => {
  const drilldownItem = component
    .dive()
    .childAt(1)
    .childAt(2);
  const { item, type, onSelect } = drilldownItem.props();
  expect(type).toEqual('switch');

  onSelect(item);
  expect(onToggle).toHaveBeenCalledWith(item);
});
