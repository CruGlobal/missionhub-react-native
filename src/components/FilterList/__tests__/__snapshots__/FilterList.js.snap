// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <Header
    left={<Connect(BackButton) />}
    shadow={true}
  />
  <ScrollView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Translate(FilterItem)
      item={
        Object {
          "id": "1",
          "text": "option1",
        }
      }
      key="1"
      onSelect={[Function]}
      type="drilldown"
    />
    <Translate(FilterItem)
      item={
        Object {
          "id": "2",
          "text": "option 2",
        }
      }
      key="2"
      onSelect={[Function]}
      type="drilldown"
    />
    <Translate(FilterItem)
      isSelected={true}
      item={
        Object {
          "id": "3",
          "selected": true,
        }
      }
      key="3"
      onSelect={[Function]}
      type="switch"
    />
    <Translate(FilterItem)
      isSelected={false}
      item={
        Object {
          "id": "2",
          "selected": false,
        }
      }
      key="2"
      onSelect={[Function]}
      type="switch"
    />
  </ScrollView>
</View>
`;
